feat: Build and publish a QEMU image artifact

darora opened the pull request:
Creates and publishes a QEMU image for our Postgres distribution.

The build process is adapted from our AMI build, but is slightly different, as described in qemu_artifact.md

soedirgo committed:
fix: account for `public` grantee

soedirgo committed:
fix(ci): respect postgresVersion input (#1237)

kangmingtay committed:
feat: bump gotrue version to v2.162.0 (#1241)

soedirgo committed:
fix: only grant pg_read_all_data if it exists (#1242)

* fix: only grant pg_read_all_data if it exists

* fix: prevent `public` from being casted into `regrole`

pcnc committed:
fix(15.6): account for pg_stat_monitor major version upgrade (#1247)

dragarcia committed:
chore: release updates to run physical backup as a service to 15.6 image (#1248)

* chore: updates to run physical backups as a service (#1235)

* chore: init commence-backup service

* chore: bump adminapi and adminmgr

* chore: bump version

* fix: provide correct filename

* chore: bump postgres version

pcnc committed:
fix(15.6/pg_upgrade): retry commands within the cleanup step; wait until PG is ready to accept connections (#1250)

pcnc committed:
fix(15.6/upgrades): collision when patching wrappers lib locations for upgrades (#1252)

hf committed:
feat: bump auth 2.162.1 on 15.6 (#1256)

pcnc committed:
fix(15.6): disable pg_stat_monitor (#1260)

* fix: disable pg_stat_monitor

* chore: bump version

pcnc committed:
fix(15.6): disable pg_stat_monitor (#1262)

* fix: disable pg_stat_monitor

* chore: bump version

kangmingtay committed:
feat: bump gotrue to v2.162.2 (#1264)

pcnc committed:
chore: add timescaledb 2.9.1; wrappers upgrade fix; wrappers & plv8 naming fix (#1259)

Co-authored-by: Bobbie Soedirgo <bobbie@soedirgo.dev>
Co-authored-by: Bobbie Soedirgo <31685197+soedirgo@users.noreply.github.com>
Co-authored-by: Kang Ming <kang.ming1996@gmail.com>
Co-authored-by: Stojan Dimitrovski <sdimitrovski@gmail.com>
Co-authored-by: Sam Rose <samuel@supabase.io>
fix(ci): respect postgresVersion input (#1237)
fix: only grant pg_read_all_data if it exists (#1242)
fix(15.6): disable pg_stat_monitor (#1260)

pcnc committed:
chore(15.6): bump pg version (#1273)

hf committed:
feat: bump auth to v2.163.0 on 15.6 (#1275)

pcnc committed:
fix: restart PG during pre-upgrade steps to shed hanging connections (#1271)

pcnc committed:
fix(upgrades): wrappers 4.2.0 -> wrappers 4.2.0 (#1278)

soedirgo committed:
fix: handle supabase_admin authenticator membership snowflake

hf committed:
feat: add auth v2.163.1 to 15.6 (#1283)

kangmingtay committed:
feat: bump gotrue to v2.163.2 (#1287)

olirice committed:
pg_net 0.11 on 15.6 release branch (#1290)

* bump pg_net to 0.11.0

* bump image to 15.6.1.135

pcnc committed:
fix: don't copy custom extension scripts during pg_upgrade (#1291)

pcnc committed:
fix: add recursive flag to custom extension script directory delete (#1292)

pcnc committed:
[GEN-11027] chore: reserve a fixed amount of blocks for the data volume; remove ansible pkg + ppa (#1295)

soedirgo committed:
fix: grant predefined roles post-upgrade

soedirgo committed:
fix: add more roles to reserved_roles & reserved_memberships (#1303)

* fix: add more roles to reserved_roles & reserved_memberships

* Update common-nix.vars.pkr.hcl

hf committed:
feat: update envoy lds config with auth jwks, oidc URLs, strip `sb-opk` header (#1296)

dragarcia committed:
chore: udpate package repo for salt (#1307)

soedirgo committed:
fix: use sudo for apt-get commands

dragarcia committed:
chore: cleanup pgbouncer.get_auth ownership accordingly

dragarcia committed:
chore: standardize

darora committed:
chore: try a qemu build

olirice committed:
Update pg net to 0.13.0 on pg 15.6 branch (#1315)

* upgrade pg_net to 0.13.0 on 15.6

* bump postgres-version

darora committed:
chore: avoid using ansible globbing

darora committed:
revert stage2 changes

darora committed:
preserve AMI build process

darora committed:
chore: check ec2ic perms if not in qemu mode

darora committed:
chore: further ami build things

darora committed:
chore: ship qcow2 image instead of raw

darora committed:
chore: cleanup + start building nix artifacts on qemu branches

darora committed:
chore: add rudimentary docs on the QEMU artifact

darora committed:
chore: upgrade to latest adminapi release

dragarcia committed:
chore: first pass at adding details to the docs

dragarcia committed:
chore: update doc on steps when running within the cluster

darora committed:
Update qemu_artifact.md

darora committed:
chore: add image built on alpine with some qemu deps

darora committed:
chore: upgrade to latest alpine

darora committed:
Merge remote-tracking branch 'origin/develop' into da/merging-qemu

darora committed:
chore: remove temporary changes that are no longer necessary

darora committed:
chore: upgrade to latest awscli release

darora committed:
fix: define postgres major version

darora committed:
fix: more new vars

darora committed:
fix: yet another pg var

darora committed:
Merge remote-tracking branch 'origin/develop' into da/merging-qemu

darora committed:
chore: initial version of GHA to build qemu artifact

darora committed:
.

darora committed:
.

darora committed:
.

darora committed:
chore: install system dependencies

darora committed:
chore: pass in pg version var correctly

darora committed:
enable logs

darora committed:
.

darora committed:
.

darora committed:
chore: temporarily disable KVM to test sw emulation performance

darora committed:
chore: specify number of CPU cores

darora committed:
trigger build

darora committed:
Revert "chore: temporarily disable KVM to test sw emulation performance"

This reverts commit 2f0f1bea8226a295bcf4fab4fd8b093e6b8d1aef.

darora committed:
chore: use native runners

darora committed:
update apt sources

darora committed:
.

darora committed:
attempt to enable kvm

darora committed:
.

darora committed:
chore; REVERT this later

darora committed:
.

darora committed:
Revert "chore; REVERT this later"

This reverts commit f9643937bbe0cea54a569a2eb1ed1ac741fee875.

darora committed:
chore: remove path output

darora committed:
.

darora committed:
chore: install packer since we're experimenting with a base image

darora committed:
reenable host cpu

darora committed:
.

darora committed:
try publishing

darora committed:
cleanup

darora committed:
.

darora committed:
fix

darora committed:
.

darora committed:
.

darora committed:
.

darora committed:
test using builtin deps

darora committed:
Revert "test using builtin deps"

This reverts commit 8d5d14a800785a81714e7e2447901158de3eceb8.

darora committed:
.

darora committed:
simplify

darora committed:
.

darora committed:
context

darora committed:
fix: correct tag

darora committed:
Merge remote-tracking branch 'origin/develop' into da/merging-qemu

darora committed:
chore: cleanup instructions for building image

darora commented:
Probably simplest to squash-merge this change

darora commented on .github/workflows/ami-release-nix.yml:
```diff
@@ -78,6 +78,7 @@ jobs:
         run: |
           packer init amazon-arm64-nix.pkr.hcl
           GIT_SHA=${{github.sha}}
+          # why is postgresql_major defined here instead of where the _three_ other postgresql_* variables are defined?
```
@samrose not sure if you know?

samrose replied:
@darora this var is used in ansible, which in turn is run by packer, and so that is the most likely reason it was set here

```
git grep postgresql_major
.github/workflows/ami-release-nix.yml:          packer build -var "git-head-version=${GIT_SHA}" -var "packer-execution-id=${GITHUB_RUN_ID}"  -var-file="development-arm.vars.pkr.hcl" -var-file="common-nix.vars.pkr.hcl" -var "ansible_arguments=-e postgresql_major=${POSTGRES_MAJOR_VERSION}"  amazon-arm64-nix.pkr.hcl
.github/workflows/testinfra-nix.yml:          packer build -var "git-head-version=${GIT_SHA}" -var "packer-execution-id=${GITHUB_RUN_ID}" -var-file="development-arm.vars.pkr.hcl" -var-file="common-nix.vars.pkr.hcl" -var "ansible_arguments=" -var "postgres-version=${{ steps.random.outputs.random_string }}" -var "region=ap-southeast-1" -var 'ami_regions=["ap-southeast-1"]' -var "force-deregister=true" -var "ansible_arguments=-e postgresql_major=${POSTGRES_MAJOR_VERSION}" amazon-arm64-nix.pkr.hcl
CONTRIBUTING.md:ADD "https://github.com/supabase/pg_graphql/releases/download/v${pg_graphql_release}/pg_graphql-v${pg_graphql_release}-pg${postgresql_major}-${TARGETARCH}-linux-gnu.deb" \
Dockerfile-15:ARG postgresql_major=15
Dockerfile-15:ARG postgresql_release=${postgresql_major}.1
Dockerfile-orioledb-17:ARG postgresql_major=17-orioledb
Dockerfile-orioledb-17:ARG postgresql_release=${postgresql_major}.1
ansible/tasks/internal/collect-pg-binaries.yml:    path: /tmp/pg_binaries/{{ postgresql_major }}/
ansible/tasks/internal/collect-pg-binaries.yml:    src: /usr/lib/postgresql/{{ postgresql_major }}/{{ item }}/
ansible/tasks/internal/collect-pg-binaries.yml:    dest: /tmp/pg_binaries/{{ postgresql_major }}/{{ item }}/
ansible/tasks/internal/collect-pg-binaries.yml:    dest: /tmp/pg_binaries/{{ postgresql_major }}/lib/
ansible/tasks/internal/collect-pg-binaries.yml:    dest: /tmp/pg_binaries/{{ postgresql_major }}/lib/
ansible/tasks/internal/collect-pg-binaries.yml:    dest: /tmp/pg_binaries/{{ postgresql_major }}/lib/libpq.so.5
ansible/tasks/internal/collect-pg-binaries.yml:    src: /usr/share/postgresql/{{ postgresql_major }}/
ansible/tasks/internal/collect-pg-binaries.yml:    dest: /tmp/pg_binaries/{{ postgresql_major }}/share/
ansible/tasks/setup-postgres.yml:    name: postgresql-{{ postgresql_major }}={{ postgresql_release }}-1.pgdg20.04+1
ansible/tasks/setup-postgres.yml:    cmd: ln -s /usr/lib/postgresql/{{ postgresql_major }}/bin /usr/lib/postgresql/bin
ansible/tasks/setup-postgrest.yml:    repo: "deb http://apt.postgresql.org/pub/repos/apt/ focal-pgdg {{ postgresql_major }}"
ansible/tasks/setup-postgrest.yml:    repo: "deb http://apt.postgresql.org/pub/repos/apt/ focal-pgdg {{ postgresql_major }}"
```

Those are all of the places where it's used in ansible.

samrose replied:
There are some parts of ansible build that use postgresql_major, and some that use the full version of postgres. If you think there is a way to refactor and change that into just using the version, I am in favor of that. 

Otherwise, this was the pattern in vars.yml


```
postgres_major:
  - "15"
  - "orioledb-17"

# Full version strings for each major version
postgres_release:
  postgresorioledb-17: "17.0.1.032-orioledb"
  postgres15: "15.8.1.036"
```

samrose replied:
Or if you just think it might be better to move that line to another spot where other vars are defined, probably it won't be an issue

darora replied:
There's probably some easy minor cleanup we can do to ensure a little bit more consistency in how they're passed in, e.g. by getting rid of `ansible_arguments` and populating that from the underlying variables...will take a look at it

darora commented on .github/workflows/qemu-image-build.yml:
```diff
@@ -0,0 +1,185 @@
+          VERSION=$(cat common-nix.vars.pkr.hcl | sed -e 's/postgres-version = "\(.*\)"/\1/g')
+          echo "version=$VERSION" >> $GITHUB_OUTPUT
+
+      # - name: Create nix flake revision tarball
```
leaving some things commented out for now - will be needed later

samrose reviewed (commented):
@darora everything looks good so far. 

It seems the `qemu-image-build.yml` only builds one of the versions of pg. But maybe you have it set up that way just to speed up development?

If there is some way I can help please let me know. I don't mind reviewing again too.


It is maybe possible we might be able to eventually refactor so that we don't need stages, etc, but it might be worth waiting until until you get through a first version of this, and then try to shift away from that.

darora commented:
Yeah - we're currently only iterating on the PG15 builds so it was simpler to constrain the build to that. I imagine it should be fairly straightforward to re-enable the PG17 build as a follow-up in the next couple of weeks.

Thanks for the review - I'll have someone on infra review and approve it as well.

thebengeu commented on ansible/files/permission_check.py:
```diff
@@ -9,18 +11,18 @@
+        {'groupname': 'plugdev', 'username': 'ubuntu'},
+        {'groupname': 'sudo', 'username': 'ubuntu'},
+        {'groupname': 'ubuntu', 'username': 'ubuntu'},
+        {'groupname': 'video', 'username': 'ubuntu'}
```
Nit: inconsistent single/double quotes and spacing? But alphabetized 💚

darora replied:
Good point - ran both this and the shell script through an autoformatter.

thebengeu commented on ebssurrogate/scripts/qemu-bootstrap-nix.sh:
```diff
@@ -0,0 +1,147 @@
+then 
+	ARCH="amd64";
+else
+        ARCH="arm64";
```
Nit: inconsistent spaces/tabs throughout the file, but don't need to bother fixing.

thebengeu commented on qemu_artifact.md:
```diff
@@ -0,0 +1,50 @@
+The current AMI process involves a few steps:
+
+1. nix package is build and published using GHA (`.github/workflows/nix-build.yml`)
+  - this builds Postgres alongwith the PG extensions we use.
```
```suggestion
  - this builds Postgres along with the PG extensions we use.
```

darora committed:
chore: run scripts through an auto-formatter

Used black for python script, shfmt for shell. Default configurations
were used for both.

darora committed:
Merge remote-tracking branch 'origin/develop' into da/merging-qemu
